packages/principled-forms/tests/field.test.ts

```diff
@@ -0,0 +1,236 @@
+import Field, { Validate, DEFAULT_MISSING_MESSAGE } from '../src/field';
+import { Invalid, Valid, Validity } from '../src/validity';
+import { minValue, minLength } from '../src/validators';
+import { Maybe } from 'true-myth';
+
+describe('`field` module', () => {
+  describe('`required` function', () => {
+    test('with no arguments', () => {
+      const validated = Field.validate(Field.required<number>());
+      expect(validated.validity.type).toBe(Validity.Type.Invalid);
+      expect((validated.validity as Invalid).reason).toBe(DEFAULT_MISSING_MESSAGE);
+    });
+
+    describe('with validators', () => {
+      test('with no value', () => {
+        const unset = Field.validate(Field.required({ validators: [minValue(10)] }));
+        expect(unset.validity.type).toBe(Validity.Type.Invalid);
+        expect((unset.validity as Invalid).reason).toBe(DEFAULT_MISSING_MESSAGE);
+      });
+
+      test('with wrong value', () => {
+        const tooYoung = Field.validate(Field.required({ value: 16, validators: [minValue(18)] }));
+        expect(tooYoung.validity.type).toBe(Validity.Type.Invalid);
+        expect((tooYoung.validity as Invalid).reason).toBe('Must be at least 18');
+      });
+
+      test('with valid value', () => {
+        const oldEnough = Field.validate(Field.required({ value: 21, validators: [minValue(18)] }));
+        expect(oldEnough.validity.type).toBe(Validity.Type.Valid);
+      });
+    });
+  });
+
+  describe('`optional` function', () => {
+    test('with no arguments', () => {
+      const f = Field.optional();
+
+      expect(Field.validate(f).validity).toBeInstanceOf(Valid);
+    });
+
+    describe('with validators', () => {
+      test('with no value', () => {
+        const unset = Field.validate(Field.optional({ validators: [minValue(10)] }));
+        expect(unset.validity.type).toBe(Validity.Type.Valid);
+      });
+
+      test('with wrong value', () => {
+        const tooYoung = Field.validate(Field.optional({ value: 16, validators: [minValue(18)] }));
+        expect(tooYoung.validity.type).toBe(Validity.Type.Invalid);
+        expect((tooYoung.validity as Invalid).reason).toBe('Must be at least 18');
+      });
+
+      test('with valid value', () => {
+        const oldEnough = Field.validate(Field.optional({ value: 21, validators: [minValue(18)] }));
+        expect(oldEnough.validity.type).toBe(Validity.Type.Valid);
+      });
+    });
+  });
+
+  describe('`validate` function', () => {
+    describe('with `Validate.Eagerly`', () => {
+      test('for `optional` fields', () => {
```

Comment:
Consider doing here what you did below: make this `describe` instead of `test` and then wrap the `expect` calls in a `test` that describes what the test is (e.g. `'no value is ``Invalid``'`).

Reply:
Great suggestion.